webapp/api/github.py

```diff
@@ -18,59 +18,121 @@ def __init__(self, access_token=None, session=api.requests.Session()):
         """
         Lists repositories that the authenticated user has explicit permission
         to access.
         """
-        repos = []
-        resp = self.get_resource(f"user/repos", page=page)
+        gql = (
+            """{
+              viewer {
+                repositories(
+                    first: 100,
+                    ownerAffiliations:
+                        [OWNER, ORGANIZATION_MEMBER],
+                """
+            + (f'after: "{end_cursor}"' if end_cursor else "")
+            + """
+                ) {
+              edges {
+                node {
+                  nameWithOwner
+                  yamlLocation1:
+                    object(expression: "master:snapcraft.yaml")
+                    { id }
+                  yamlLocation2:
+                    object(expression: "master:.snapcraft.yaml")
+                    { id }
+                  yamlLocation3:
+                    object(expression: "master:snap/snapcraft.yaml")
+                    { id }
+                  yamlLocation4:
+                    object(expression: "master:build-aux/snap/snapcraft.yaml")
+                    { id }
+                }
+              }
+              pageInfo {
+                hasNextPage
+                endCursor
+              }
+            }
+          }
+        }"""
+        )
+
+        gql_response = self._request(gql)["viewer"]["repositories"]
+        page_info = gql_response["pageInfo"]
+        repositories = gql_response["edges"]
+        processed_repos = []
 
-        if resp.ok:
-            repos.extend(resp.json())
+        for repo in repositories:
+            processed_repo = {}
 
-        if all_pages and "next" in resp.links:
-            repos.extend(self.get_user_repositories(page=page + 1))
+            processed_repo["name"] = repo["node"]["nameWithOwner"]
+            processed_repo["snapcraft_yaml"] = any(
+                [bool(repo["node"][f"yamlLocation{i}"]) for i in range(1, 4)]
+            )
+            processed_repos.append(processed_repo)
 
-        return repos
+        if page_info["hasNextPage"]:
+            processed_repos.extend(
+                self.get_user_repositories(page_info["endCursor"])
+            )
 
-    def get_user_orgs(self, all_pages=True, page=1):
+        return processed_repos
+
+    def is_snapcraft_yaml_present(self, owner, repo):
         """
-        Lists organizations that the authenticated user has explicit permission
-        to access.
+        Check is the snapcraft.yaml file exist in the GitHub repo
         """
-        orgs = []
-        resp = self.get_resource(f"user/orgs", page=page)
-
-        if resp.ok:
-            orgs.extend(resp.json())
+        gql = (
+            '''{
+            repository(name: "'''
+            + repo
+            + '''", owner: "'''
+            + owner
+            + """") {
+                id
+                nameWithOwner
+                yamlLocation1:
```

Comment:
is there a way to template some of this out with GraphQL? For instance the repetition of these yamlLocations?

Reply:
Luke I changed this request to use the Rest API because we might don't need the user session to verify this.